File: leetcode/leetcode_cpp/sum-root-to-leaf-numbers.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
 /*
 https://leetcode.com/problems/sum-root-to-leaf-numbers
 
 Solution1-2. DFS / BFS
 time: o(n)
 space: o(n)
 */
class Solution {
public:
    int sumNumbers_DFS(TreeNode* node, int cur) {
        if (!node) return 0;
        if (!node->left && !node->right) return cur + node->val;

        cur = (cur + node->val) * 10;
        int val = sumNumbers_DFS(node->left, cur);
        val += sumNumbers_DFS(node->right, cur);
        return val;
    }

    int sumNumbers_1(TreeNode* root) {
        return sumNumbers_DFS(root, 0);
    }

    int sumNumbers_2(TreeNode* root) {
        queue<pair<TreeNode*, int>> q;
        q.push({root, 0});

        int ans = 0;
        while (!q.empty()) {            
            auto value = q.front(); q.pop();
            auto node = value.first;
            auto val = value.second;
            if (!node) continue;

            val = (val * 10) + node->val;
            if (!node->left && !node->right) {
                ans += val;
            }
            if (node->left) q.push({node->left, val});
            if (node->right) q.push({node->right, val});
        }
        return ans;
    }

    int sumNumbers(TreeNode* root) {
        return sumNumbers_2(root);
    }
};

File: leetcode/leetcode_py/sum-root-to-leaf-numbers.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
'''
https://leetcode.com/problems/sum-root-to-leaf-numbers
 
Solution1-2. DFS / BFS
time: o(n)
space: o(n)
'''
class Solution:
    def sumNumbers_1(self, root: Optional[TreeNode]) -> int:
        def dfs(node, cur):
            if not node: return 0
            if not node.left and not node.right: return cur + node.val

            cur = (cur + node.val) * 10
            val = dfs(node.left, cur)
            val += dfs(node.right, cur)
            return val
        return dfs(root, 0)

    def sumNumbers_2(self, root: Optional[TreeNode]) -> int:
        q = []
        q.append((root, 0))

        ans = 0
        while q:
            node, val = q.pop(0)
            if not node: continue
            if not node.left and not node.right:
                ans += val + node.val
                continue
            val = (val + node.val) * 10
            if node.left:
                q.append((node.left, val))
            if node.right:
                q.append((node.right, val))

        return ans
    def sumNumbers(self, root: Optional[TreeNode]) -> int:
        return self.sumNumbers_2(root)
        